### PR TITLE
Fixes for issues #1 and #2

### PR DESCRIPTION
**Pull request by pfumagalli**
This fixes killbill/killbill-plugin-framework-java#1 and killbill/killbill-plugin-framework-java#2

### CODE REVIEW

**Comment by pfumagalli** [comment]
I've added another patch that strips out the known properties from "additional data".

As per the example in the Adyen plugin, those values could be retrieved from the record stored in the database (like [here](https://github.com/killbill/killbill-adyen-plugin/blob/master/src/main/java/org/killbill/billing/plugin/adyen/api/AdyenPaymentPluginApi.java#L441-L454)).

It's somewhat scary to be returned CC number and verification code by default...

File: src/main/java/org/killbill/billing/plugin/dao/payment/PluginPaymentDao.java [hunk 1]
```diff
@@ -159,9 +160,57 @@ public RESP_R withConnection(final Connection conn) throws SQLException {
...
+        final String ccTrackData         = getProperty(PluginPaymentPluginApi.PROPERTY_CC_TRACK_DATA,         clonedProperties);
+        final String address1            = getProperty(PluginPaymentPluginApi.PROPERTY_ADDRESS1,              clonedProperties);
+        final String address2            = getProperty(PluginPaymentPluginApi.PROPERTY_ADDRESS2,              clonedProperties);
+        final String city                = getProperty(PluginPaymentPluginApi.PROPERTY_CITY,                  clonedProperties);
+        final String state               = getProperty(PluginPaymentPluginApi.PROPERTY_STATE,                 clonedProperties);
+        final String zip                 = getProperty(PluginPaymentPluginApi.PROPERTY_ZIP,                   clonedProperties);
+        final String country             = getProperty(PluginPaymentPluginApi.PROPERTY_COUNTRY,               clonedProperties);
+
+        /* Remove the keys from the map that will become the "additional data" */
+        clonedProperties.remove(PluginPaymentPluginApi.PROPERTY_TOKEN);
```
Comment (pierre) [inline src/main/java/org/killbill/billing/plugin/dao/payment/PluginPaymentDao.java]: Could it be simplified (to make sure we don't forget to remove some if we add new properties down the line)?

Maybe:

```
final String token = Strings.emptyToNull(clonedProperties.remove(PluginPaymentPluginApi.PROPERTY_TOKEN));
```

(we could even create a method for it like `popProperty`)

Comment (pfumagalli) [inline src/main/java/org/killbill/billing/plugin/dao/payment/PluginPaymentDao.java]: Hmm... `Map.remove(key)` already removes the key and returns the value that was associated with it (or `null`), so I'd say it's just a matter of doing something like:

```java
final String token = clonedProperties.remove(PluginPaymentPluginApi.PROPERTY_TOKEN);
```

**Comment by pierre** [comment]
Thanks! It looks good overall, comments inline.

> I've added another patch that strips out the known properties from "additional data".

That's fine. Could you add a test for it?

> It's somewhat scary to be returned CC number and verification code by default...

Kill Bill plugins cannot handle real CC numbers at the moment, since they are stored in clear (see also section _How can I secure my Kill Bill installation?_ of http://killbill.io/faq/). It is expected that the `ccNumber` attribute is a token (potentially passing the luhn check).

That being said, we could add support for it, but it involves some work in the DAO to support symmetric encryption and a mechanism to easily rotate encryption keys. CVV should also not be stored.

**Comment by pfumagalli** [comment]
On Tue, Feb 17, 2015 at 12:58 AM, Pierre-Alexandre Meyer
<notifications@github.com> wrote:
>
> Thanks! It looks good overall, comments inline.
>
>> I've added another patch that strips out the known properties from "additional data".
>>
> That's fine. Could you add a test for it?
>
There are no tests running on `killbill-plugin-framework`, shall I start?

>> It's somewhat scary to be returned CC number and verification code by default...
>>
> Kill Bill plugins cannot handle real CC numbers at the moment, since they are stored in clear (see also section How can I secure my Kill Bill installation? of http://killbill.io/faq/). It is expected that the ccNumber attribute is a token (potentially passing the luhn check).
>
> That being said, we could add support for it, but it involves some work in the DAO to support symmetric encryption and a mechanism to easily rotate encryption keys. CVV should also not be stored.

Absolutely agreed, but I still think that those properties should
*NOT* be duplicated in the column *AND* additional data in the base
plugin, because someone might take all the various precautions about
the table itself, and magically forget about the JSON.

**Comment by pierre** [comment]
> There are no tests running on `killbill-plugin-framework`, shall I start?

That'd be great! For the history, this repo is brand new and was extracted after working on the [Adyen](https://github.com/killbill/killbill-adyen-plugin) and the [Forte](https://github.com/killbill/killbill-forte-plugin) plugins. Tests were on my TODO list...

If it helps, I've imported `TestWithEmbeddedDBBase` which should give you the plumbing to embed MySQL (the Adyen plugin has examples on how it works).

Also, FYI, the `killbill-oss-parent` pom expects `slow` and `fast` TestNG tests groups. The former are only run when the `mysql` profile is active (https://github.com/killbill/killbill-oss-parent/blob/master/pom.xml#L1599).

> Absolutely agreed, but I still think that those properties should *NOT* be duplicated in the column *AND* additional data in the base plugin, because someone might take all the various precautions about the table itself, and magically forget about the JSON.

:+1:

**Comment by pfumagalli** [comment]
I'll see what I can do on those tests...

As you know (our current common employer) needs to get Japan ASAP on
KillBill and all my time is actually spent on bringing the Japan-specific
providers plugin working!

There isn't much in the plugin framework allowing me to get started, but
I'll give it a shot :-D

**Comment by pfumagalli** [comment]
There... Now you have tests, too :+1:

**Comment by pfumagalli** [comment]
Pierre, when do you think you can merge this in and re-release?

**Comment by pierre** [comment]
Thanks for the tests! :beers:  0.0.3 is on its way to Maven Central.

FYI, I removed the diamond operators so it still compiles on 1.6.

**Pull request merged**